include/seastar/rpc/rpc_impl.hh

```diff
@@ -345,25 +345,86 @@ struct unmarshal_one {
     };
 };
 
```

Comment:
> We notice that the first value can in fact be deserialized in the tuple constructor arguments, since there's no ordering problem for it. So we remove the std::optional wrapper for it unconditionally.

I do not understand what ordering problem has to do with default constructability.   Can you elaborate?

Reply:
The non-first elements cannot be deserialized in the constructor since they suffer from the ordering problem; therefore they have to be default constructible (either naturally or by wrapping with std::optional). The first element has no ordering problem, so it can be deserialized in the constructor, so it need not be default constructible.

Reply:
Why the first element does not have ordering problem. May be I misunderstand what the "ordering problem" is. Can you define it?

Reply:
Deserializing element n+1 must happen after element n.

C++ doesn't define the order of argument evaluation, so we can't deserialize element n+1 in the constructor. So element n+1 must be default constructible. 

We can however deserialize element n, for the case n=0, so we can drop the default constructible requirement for that element.

Reply:
> Deserializing element n+1 must happen after element n.

OK. So I do understand the problem correctly. But why the first element does not suffer from it. You do not want to deserialize index 1 before index 0, no?

> 
> C++ doesn't define the order of argument evaluation, so we can't deserialize element n+1 in the constructor. So element n+1 must be default constructible.
> 
> We can however deserialize element n, for the case n=0, so we can drop the default constructible requirement for that element.

Ah, so first element is special because you want to optimize for the case where there is only one element?

Reply:
> > Deserializing element n+1 must happen after element n.
> 
> OK. So I do understand the problem correctly. But why the first element does not suffer from it. You do not want to deserialize index 1 before index 0, no?


I deserialze 0 in the constructor, and 1... in the comma expression. 

> 
> > 
> > C++ doesn't define the order of argument evaluation, so we can't deserialize element n+1 in the constructor. So element n+1 must be default constructible.
> > 
> > We can however deserialize element n, for the case n=0, so we can drop the default constructible requirement for that element.
> 
> Ah, so first element is special because you want to optimize for the case where there is only one element?

I want to remove std::optional wrappers. The first element can lose the wrapper unconditionally, the rest only if they are default constructible.

Reply:
> > > Deserializing element n+1 must happen after element n.
> > 
> > 
> > OK. So I do understand the problem correctly. But why the first element does not suffer from it. You do not want to deserialize index 1 before index 0, no?
> 
> I deserialze 0 in the constructor, and 1... in the comma expression.
> 

Ah, finally I get it. Since you de-serialize only one the order does not matter.